File: src/2023/01/AoC_2023_01.py

def get_lines_from_file_name(file_name):
    data = open(file_name, 'r')
    return data.readlines()

def get_first_number(line):
    for i, a in enumerate(line):
        match a:
            case "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9":
                return a
            case "z":
                if "zero" == line[i: i+len("zero")]:
                    return "0"
            case "o":
                if "one" == line[i: i+len("one")]:
                    return "1"
            case "t":
                if "two" == line[i: i+len("two")]:
                    return "2"
                if "three" == line[i: i+len("three")]:
                    return "3"
            case "f":
                if "four" == line[i: i+len("four")]:
                    return "4"
                if "five" == line[i: i+len("five")]:
                    return "5"
            case "s":
                if "six" == line[i: i+len("six")]:
                    return "6"
                if "seven" == line[i: i+len("seven")]:
                    return "7"
            case "e":
                if "eight" == line[i: i+len("eight")]:
                    return "8"
            case "n":
                if "nine" == line[i: i+len("nine")]:
                    return "9"

def get_last_number(line):
    # line = the original line
    # i    = index of the original line
    # line_reversed = line in reversed order
    # i_reversed    = index of the reversed line

    line_reversed = ''.join(reversed(line))
    for i_reversed, a in enumerate(line_reversed):
        i = len(line) - i_reversed - 1;
        match a:
            case "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9":
                return a
            case "z":
                if "zero" == line[i: i+len("zero")]:
                    return "0"
            case "o":
                if "one" == line[i: i+len("one")]:
                    return "1"
            case "t":
                if "two" == line[i: i+len("two")]:
                    return "2"
                if "three" == line[i: i+len("three")]:
                    return "3"
            case "f":
                if "four" == line[i: i+len("four")]:
                    return "4"
                if "five" == line[i: i+len("five")]:
                    return "5"
            case "s":
                if "six" == line[i: i+len("six")]:
                    return "6"
                if "seven" == line[i: i+len("seven")]:
                    return "7"
            case "e":
                if "eight" == line[i: i+len("eight")]:
                    return "8"
            case "n":
                if "nine" == line[i: i+len("nine")]:
                    return "9"


lines = get_lines_from_file_name("./data.txt")

result = 0
for line in lines:
    n1 = get_first_number(line)
    n2 = get_last_number(line)
    
    result += int(n1+n2)


print(f"{result}")

File: src/2023/01/AoC_2023_01.rs
//use std::error::Error;
use std::fs::File;
use std::io::BufRead;
use std::io::BufReader;
use std::io::Lines;
use std::path::Path;

fn main () {
    let lines = get_lines_from_file_name("./data.txt");

    let mut result: i32 = 0;
    for line in lines {
        match line {
            Ok(line) => {
                //println!("{}", line);
                let a1 = get_first_number(&line);
                let a2 = get_last_number(&line);

                result += (a1 * 10) + a2;
                //println!("  {} <-- adding {}", result, (a1 * 10) + a2);
            },
            Err(e) => println!("ERROR: {}", e),
        }
    }

    println!("Result: {}", result);
}


// 01 Utilities ------------------------------
fn get_first_number(line: &str) -> i32 {
    for a in line.chars() {
        match a {
            '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' => {
                //println!("  First number: {}", a);
                return a as i32 - 0x30; // https://stackoverflow.com/a/66590374
            },
            _ => { }
        }
    }
    return -1;
}
fn get_last_number(line: &str) -> i32 {
    for a in line.chars().rev() {
        match a {
            '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' => {
                //println!("  Last number {}", a);
                return a as i32 - 0x30; // https://stackoverflow.com/a/66590374
            },
            _ => { }
        }
    }
    return -1;
}


// Common Utilities ------------------------------

fn get_lines_from_file_name(file_name: &str) -> Lines<BufReader<File>> {
    let path = Path::new(file_name);
    let display = path.display();
    let file = match File::open(&path) {
        Err(why) => panic!("couldn't open {}: {}", display, why),
        Ok(file) => file,
    };
    let reader = BufReader::new(file);
    let lines = reader.lines();
    return lines;
}

// https://stackoverflow.com/a/58119924
// fn print_type_of<T>(_: &T) {
//     println!("{}", std::any::type_name::<T>())
// }